core/Enclave/enclave_hsm.cpp

```diff
@@ -202,6 +235,227 @@ sgx_status_t enclave_create_key(ehsm_keyblob_t *cmk, size_t cmk_size)
+
+    if (!BN_set_word(e, (BN_ULONG)RSA_F4))
+        goto out;
+
+    rsa_keypair = RSA_new();
+    if (rsa_keypair == NULL)
+        goto out;
+
+    switch (keyspec)
+    {
+    case EH_RSA_2048:
+        RSA_generate_key_ex(rsa_keypair, RSA_2048_KEY_BITS, e, NULL);
+        break;
+    case EH_RSA_3072:
+        RSA_generate_key_ex(rsa_keypair, RSA_3072_KEY_BITS, e, NULL);
+        break;
+    case EH_RSA_4096:
+        RSA_generate_key_ex(rsa_keypair, RSA_4096_KEY_BITS, e, NULL);
+        break;
+    default:
+        ret = SGX_ERROR_INVALID_PARAMETER;
+        goto out;
+    }
+
+    bio = BIO_new(BIO_s_mem());
+    if (bio == NULL)
+        goto out;
+
+    if (!PEM_write_bio_RSAPublicKey(bio, rsa_keypair))
+        goto out;
+
+    // pend pubkey
+    if (pubkey->datalen == 0)
+    {
+        key_size = BIO_pending(bio);
+        if (key_size < 0)
+            goto out;
+
+        pubkey->datalen = key_size;
+
+        ret = SGX_SUCCESS;
+        goto out;
+    }
+
+    // pend private key
+    if (!PEM_write_bio_RSAPrivateKey(bio, rsa_keypair, NULL, NULL, 0, NULL, NULL))
+        goto out;
+
+    key_size = BIO_pending(bio);
+    if (key_size <= 0)
+        goto out;
+
+    pem_keypair = (uint8_t *)malloc(key_size);
+    if (pem_keypair == NULL)
+        goto out;
+
+    if (BIO_read(bio, pem_keypair, key_size) < 0)
+        goto out;
+
+    memcpy_s(pubkey->data, pubkey->datalen, pem_keypair, pubkey->datalen);
+
+    ret = ehsm_create_keyblob(pem_keypair, key_size, (sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    if (ret != SGX_SUCCESS)
+        goto out;
+
+out:
+    if (rsa_keypair)
+        RSA_free(rsa_keypair);
+    if (bio)
+        BIO_free(bio);
+    if (e)
+        BN_free(e);
+
+    SAFE_MEMSET(pem_keypair, key_size, 0, key_size);
+    SAFE_FREE(pem_keypair);
+
+    return ret;
+}
+
+sgx_status_t enclave_import_key_material(ehsm_keyblob_t *cmk, size_t cmk_size,
+                                         ehsm_padding_mode_t padding_mode,
+                                         ehsm_data_t *key_material, size_t key_material_size)
+{
+    sgx_status_t ret = SGX_ERROR_UNEXPECTED;
+
+    if (cmk == NULL ||
+        cmk_size != APPEND_SIZE_TO_KEYBLOB_T(cmk->keybloblen) ||
+        cmk->metadata.origin != EH_EXTERNAL_KEY ||
+        key_material_size != APPEND_SIZE_TO_DATA_T(key_material->datalen) ||
+        (cmk->metadata.keyusage != EH_KEYUSAGE_ENCRYPT_DECRYPT && cmk->metadata.keyusage != EH_KEYUSAGE_SIGN_VERIFY))
+    {
+        return SGX_ERROR_INVALID_PARAMETER;
+    }
+
+    ehsm_data_t *import_key = NULL;
+    int retval = 0;
+
+    int paddingMode = getPaddingMode(padding_mode);
+    if (paddingMode != RSA_PKCS1_PADDING && paddingMode != RSA_PKCS1_OAEP_PADDING)
+        return SGX_ERROR_INVALID_PARAMETER;
+
+    uint8_t *rsa_keypair = NULL;
+    BIO *bio = NULL;
+    RSA *rsa_prikey = NULL;
+
+    uint32_t keylength = ehsm_get_gcm_ciphertext_size((sgx_aes_gcm_data_ex_t *)cmk->keyblob);
+    switch (keylength)
```

Comment:
what length is expected? 
need to check it's not UINT32_MAX

Reply:
I just deleted it